### PR TITLE
enable AtomicReference everywhere

### LINKED ISSUE #64: Atomics cannot be used in a Swift Package Index package with AtomicReference enabled.
see https://forums.swift.org/t/swift-atomics-no-way-to-enable-atomicreference-on-linux/63169 for a more detailed explanation.

currently it is not possible to publish a valid package on the swiftpackageindex.com that depends on `swift-atomics` with `AtomicReference` enabled, because doing so requires unsafe build flags. there is no reason to require these build flags to be passed, because the swift runtime now guarantees support for doublewide compare-and-exchange anyway.

it is currently not possible to work-around the issue by patching and vendoring the `swift-atomics` libraries, because `_AtomicShims` is a C target, and cannot be aliased to a different name. therefore, the patched `_AtomicShims` module will always conflict with the original module during compilation. this conflict occurs because other packages, such as `swift-nio`, also depend on `swift-atomics`.

if `swift-atomics` were to enable `AtomicReference` by default and tag a minor release in the canonical `apple/swift-atomics` repository, libraries that depend on both `swift-atomics` and `swift-nio` would be instantly unblocked, because `swift-nio` specifies an [open `swift-atomics` requirement](https://github.com/apple/swift-nio/blob/main/Package.swift#L142).

### PR DESCRIPTION
**Pull request by kelvin13**
fixes #64 , and updates documentation to remove references to custom build flags.

<s>i was not able to build the tests in their entirety, the 40k gybbed `Basics.swift` overwhelms my local toolchain.</s> but i tried to make individual subsets of `Basics.swift` compile a little faster (~15%) by adding type annotations.

side note:

compilation time seems to grow linearly with the number of types gyb generates tests for.

| n | time |
| --- | --- |
| 1 | 4.5s |
| 2 | 10s |
| 3 | 20s |
| 4 | 29s |
| 5 | 40s |

the smaller time for the first *n* is probably because `Int` benefits from being the default inferred integer literal type.

```python
  types = [
    ("Int",        "Int",    "12", "23"),
    ("Int8",       "Int8",   "12", "23"),
    ("Int16",      "Int16",  "12", "23"),
    ("Int32",      "Int32",  "12", "23"),
    ("Int64",      "Int64",  "12", "23"),
    ("UInt",       "UInt",   "12", "23"),
    ("UInt8",      "UInt8",  "12", "23"),
# ...
```

### CODE REVIEW

**Comment by kelvin13** [comment]
update: i was able to build and run the tests and verified they all pass :)

**Comment by stephentyrone** [comment]
My gut feeling is that we should resolve this by unconditionally defining ENABLE_DOUBLEWIDE_ATOMICS rather than removing the checks, to make supporting other platforms that don't have support easier in the future.

**Comment by kelvin13** [comment]
i don't have a strong opinion on this, i would just prefer some version of this to be merged and tagged as soon as possible to unblock myself.

@lorentey which approach do you prefer?

**Comment by kelvin13** [comment]
> My gut feeling is that we should resolve this by unconditionally defining ENABLE_DOUBLEWIDE_ATOMICS rather than removing the checks, to make supporting other platforms that don't have support easier in the future. Otherwise someone is going to port Swift to some constrained environment and they'll have to reintroduce all the checks.

i might just be brainfarting here, but how do you unconditionally define `ENABLE_DOUBLEWIDE_ATOMICS` within a swift source module? i don’t think i have ever used the compiler directives in this way.

**Comment by kelvin13** [comment]
according to https://forums.swift.org/t/how-to-use-compiler-directives-to-define-a-build-flag/63191/4 there is no way to accomplish this in swift today. @stephentyrone do you know of an alternative approach we can take here?

**Comment by lorentey** [comment]
Unconditionally enabling double-wide atomics is the right move. It'll need to be tied to a release that bumps the minimum toolchain version to a Swift version whose runtime also requires 16-byte atomics, but other than that this should be smooth.

**Comment by lorentey** [comment]
(I wouldn't like to leave the conditionals in place, unless we currently still support a platform that doesn't have wide atomics.)

**Comment by kelvin13** [comment]
great! this is what the PR implements right now. anything you would like changed, or is this good to merge?

File: Utilities/run-full-tests.sh [hunk 1]
```diff
@@ -69,7 +69,7 @@ try() {
 try "spm.debug.build" $swift build -c debug $spm_flags --build-path "$build_dir/spm.debug"
 try "spm.release.build" $swift build -c release $spm_flags --build-path "$build_dir/spm.release"
 if [ "$(uname)" = "Linux" ]; then
-    try "spm.release.dword.build" $swift build -c release $spm_flags -Xcc -mcx16 -Xswiftc -DENABLE_DOUBLEWIDE_ATOMICS --build-path "$build_dir/spm.release.dword"
+    try "spm.release.dword.build" $swift build -c release $spm_flags --build-path "$build_dir/spm.release.dword"
 fi
```
Comment (lorentey) [inline Utilities/run-full-tests.sh:73]: Nit: we should just remove this entire `if` altogether -- this configuration is now the default, so it is rolled into the default Linux case.

File: Utilities/run-full-tests.sh [hunk 2]
```diff
@@ -133,7 +133,7 @@ try "spm.debug.test" $swift test -c debug $spm_flags --build-path "$build_dir/sp
 try "spm.release.test" $swift test -c release $spm_flags --build-path "$build_dir/spm.release"
 
 if [ "$(uname)" != "Darwin" ]; then
-    try "spm.release.dword.test" $swift test -c release $spm_flags -Xcc -mcx16 -Xswiftc -DENABLE_DOUBLEWIDE_ATOMICS --build-path "$build_dir/spm.release.dword"
+    try "spm.release.dword.test" $swift test -c release $spm_flags --build-path "$build_dir/spm.release.dword"
 fi
```
Comment (lorentey) [inline Utilities/run-full-tests.sh:137]: Same here -- let's remove this `if`.

**Review by lorentey — APPROVED** [review]
Looks good. We'll also need to bump the required toolchain version in the package manifest.

**Comment by lorentey** [comment]
@swift-ci test

**Comment by lorentey** [comment]
@swift-ci test

**Comment by kelvin13** [comment]
YESS

**Pull request merged**